```
[general]
# General parameters for the organization and management of the data
path = string # Path where project_name folder is or will be created
ploidy = 1 | 2 # Values that represent the ploidy of the generated individuals. If not presence or left blank, default value = 1.
project_name = string # name of the project.
output_folder_name | ofn = string  # Name of the output folder that will be created under path/project_name


[data]
# Block that is specific to the data origin and extra files needed
origin = 1 | 2 | [3]
# (1) SimPhy project output
# (2) INDELible + tree
# [3] INDELible + tree + reference sequence

filter_simphy = on - 1 - true | off - 0 - false (1)
# when ploidy do not match the number of even individuals per taxa of the simphy project
# whether to filter the replicates or not. If false, and condition is not satisfied, run fill terminate.
data_prefix | dp (1) # prefix of the FASTA files generated by simphy
indelible_control (2) # Modified INDELible control file
newick_file (2) # path of the file where the newick tree is going to be stored. Must have 1 tree.
reference_sequence [3] # File path for the FASTA file that contains the reference sequence that will be used to re-root the tree
and modified according to the parameters specified in the INDELible control file.
tip_label  [3] # Label of the tip that correspond to the reference sequence.

[coverage]
# If simulating data from targeted sequencing experiments
on_target # percentage of loci that will be considered to be on target. Optional. Default=100
off_target  # percentage of loci that will be considered to be off target. Optional. Default = 0
not_captured
# percentage of loci that will be considered as not captured. Coverage = 0x. Optional. Default = 0.
# If on/off target parameters set, and sum not not equal to 1. Value = 1 - (on_target + off_target)

experiment = # Expected coverage for experiment. Follows distribution.
# Each value sampled will be used for base coverage of a replicate.
locus = # Follows distribution. Parameter that will described shape of a Gamma distribution with mean 1. Used as a multiplier locus-wide.  
# Parameters for the alpha shape are sampled per replicate. Gamma is sampled per number of loci.
individual  # Follows distribution. Parameter that will described shape of a Gamma distribution with mean 1. Used as a multiplier individual-wide.  
# Parameters for the alpha shape are sampled per replicate. Gamma is sampled per number of individuals.

phylogenetic decay = +[ID,COV_PER]
# (Species ID, Percentage of coverage related to expected) (If used, at least 1	)

[ngs-read-counts]
error = # Sequencing error rate for the read counts. Fixed for all variable sites.
reference = # Label of the tip that will be used as reference to describe the variable sites.
# If 1 given, all replicates used will have the same reference.
# If there are several replicates, and reference vary per replicate, a file must be given.
# If any of the replicates do not have such tip, tip 1_0_0 will be used as a reference by default.
# If there is no reference label,  tip 1_0_0 will be used as a reference by default.

[ngs-reads-art]
all art parameters but not I/O and coverage related.

[execution]
environment = bash | sge | slurm
run = on - 1 - true | off - 0 - false
threads = 1 (default)
```
